## Exploring and cleaning eBay Car Sales Data

In this project we take a look at a used cars dataset from ebay, which was uploaded by user Orgesleka. The data set samples 50 000 values and is provided with a dictionary, which contains several quatitative as well as, qualitative variables.The aim of this project is to clean and analyze the data in order to extract some interesting findings.

In [2]:
import pandas as pd
import numpy as np

In [3]:
autos = pd.read_csv('autos.csv', encoding = "Latin-1")

In [4]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [5]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Using the DataFrame.info() code we can get a better look at the dataset. There are 15 object type variables and 4 integer types. However, looking at some of the object column titles, it might be the case that some were mislabelled as objects and will need to be converted to int types. Additionally, the The following rows all had roughly 3000 missing values: "gearbox", " vehicleType", "model", "fueltype", and "notrepaireddamage".  

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

The changes made to the column names included:
- converting from camelcase to snakecase
- changing the wording of some columns to make them more intuitive

## Autos cleaning and Exploration

In [8]:
autos.describe(include ='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-21 20:37:19  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

when looking at the descriptive statistics for both numerical and string based columns we can identify a few outliers that need further attention. For example, the numerical column "num_photos" has the value 0 for almost all statistic entries. It could be beneficial to drop this cell as it does not provide any useful information about the dataset. The following will be a list of things that need investigation:
- The date columns need to be converted to datetime objects
- the odometer and price columns needs to be cleaned and converted to int type
- Postal codes need to be converted to objct type as their values hold no intrinsic meaning.

First, we will convert the "Price" and "Odometer" column to numeric type.


In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [10]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","")
autos["price"] = autos["price"].astype(int)

autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km","")
autos["odometer"] = autos["odometer"].astype(int)


In [11]:
autos.rename({"odometer":"odometer_km"}, axis = 1, inplace = True)

In [12]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price  ab_test vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere       150000                   3       lpg   
1      automatik       286     7er       150000                   6    benzin   
2        manuell       102    golf        70000                   7    benzin   
3      automatik        71  fortwo        70000                   6    benzin   
4        manuell         0   focus       150000                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5       100000                   1    diesel   
49996    manuell        75   astra       150000                   5    benzin   
49997  automatik        69     500         5000                  11    benzin   
49998    manuell       150      a3        40000                  11    diesel   
49999    manuell       101  vectra       150000                   1    benzin   

            brand unrepaired_damage           ad_created  num_photos  \
0         peugeot              nein  2016-03-26 00:00:00           0   
1             bmw              nein  2016-04-04 00:00:00           0   
2      volkswagen              nein  2016-03-26 00:00:00           0   
3           smart              nein  2016-03-12 00:00:00           0   
4            ford              nein  2016-04-01 00:00:00           0   
...           ...               ...                  ...         ...   
49995        audi              nein  2016-03-27 00:00:00           0   
49996        opel              nein  2016-03-28 00:00:00           0   
49997        fiat              nein  2016-04-02 00:00:00           0   
49998        audi              nein  2016-03-08 00:00:00           0   
49999        opel              nein  2016-03-13 00:00:00           0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  2016-04-01 14:38:50  
...            ...                  ..

In [13]:
#autos["price"].unique().shape
#autos["price"].describe()
autos_price_counts = autos["price"].value_counts()
autos_price_counts.sort_index(ascending=True).head(50)



0      1421
1       156
2         3
3         1
5         2
8         1
9         1
10        7
11        2
12        3
13        2
14        1
15        2
17        3
18        1
20        4
25        5
29        1
30        7
35        1
40        6
45        4
47        1
49        4
50       49
55        2
59        1
60        9
65        5
66        1
70       10
75        5
79        1
80       15
89        1
90        5
99       19
100     134
110       3
111       2
115       2
117       1
120      39
122       1
125       8
129       1
130      15
135       1
139       1
140       9
Name: price, dtype: int64

In [14]:
# for the higher end of the prices spectrum, cut off anything 10000000 and over 
# for the lower end of the price spectrum, remove anything 100 or under

using the value_count and describe method on the price row, we discover that several thousand cars had non-intuitive prices. For example, some cars were going for under 100 while other were exceediing 10 million. By removing these entries, we can control for more realistic values

In [15]:
autos_odometer_counts = autos["odometer_km"].value_counts()
autos_odometer_counts.sort_index(ascending=True).head(50)


5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [16]:
autos = autos[autos["price"].between(100,10000000)]


In [17]:
autos["odometer_km"].max()
autos["price"].value_counts()

500      781
1500     734
2500     643
1000     639
1200     639
        ... 
846        1
2895       1
41850      1
910        1
29890      1
Name: price, Length: 2316, dtype: int64

the remaining values for price and odometer_km reflect a realistic and traditional entries for ebay. Most vehicles being sold exceed the 150000 km mark on the odometer and prices do resemble that. 

# Exploring the date columns

In [18]:
autos["date_crawled"].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025357
2016-03-06    0.014037
2016-03-07    0.036056
2016-03-08    0.033174
2016-03-09    0.033008
2016-03-10    0.032282
2016-03-11    0.032593
2016-03-12    0.036906
2016-03-13    0.015675
2016-03-14    0.036657
2016-03-15    0.034314
2016-03-16    0.029462
2016-03-17    0.031515
2016-03-18    0.012896
2016-03-19    0.034729
2016-03-20    0.037797
2016-03-21    0.037217
2016-03-22    0.032904
2016-03-23    0.032282
2016-03-24    0.029442
2016-03-25    0.031494
2016-03-26    0.032303
2016-03-27    0.031121
2016-03-28    0.034957
2016-03-29    0.034148
2016-03-30    0.033733
2016-03-31    0.031847
2016-04-01    0.033692
2016-04-02    0.035600
2016-04-03    0.038606
2016-04-04    0.036574
2016-04-05    0.013062
2016-04-06    0.003172
2016-04-07    0.001389
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [19]:
autos["ad_created"].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036926
2016-04-05    0.011797
2016-04-06    0.003255
2016-04-07    0.001244
Name: ad_created, Length: 76, dtype: float64

In [20]:
autos["last_seen"].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001078
2016-03-06    0.004313
2016-03-07    0.005432
2016-03-08    0.007319
2016-03-09    0.009579
2016-03-10    0.010636
2016-03-11    0.012399
2016-03-12    0.023781
2016-03-13    0.008874
2016-03-14    0.012627
2016-03-15    0.015861
2016-03-16    0.016442
2016-03-17    0.028094
2016-03-18    0.007319
2016-03-19    0.015758
2016-03-20    0.020651
2016-03-21    0.020547
2016-03-22    0.021356
2016-03-23    0.018577
2016-03-24    0.019759
2016-03-25    0.019096
2016-03-26    0.016670
2016-03-27    0.015550
2016-03-28    0.020858
2016-03-29    0.022309
2016-03-30    0.024694
2016-03-31    0.023823
2016-04-01    0.022848
2016-04-02    0.024901
2016-04-03    0.025129
2016-04-04    0.024528
2016-04-05    0.125086
2016-04-06    0.221973
2016-04-07    0.132135
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [21]:
autos["registration_year"].describe()

count    48231.000000
mean      2004.728867
std         87.891383
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The "registration_year" column decsribes the year in which in the car was registered, allowing us to know its 'age'. The min value "1000" is a year when cars were not invented yet and the max "9999" holds no intuitive meaning.

In [22]:
autos["registration_year"].value_counts().sort_index()

1000    1
1001    1
1111    1
1800    2
1910    2
       ..
5911    1
6200    1
8888    1
9000    1
9999    3
Name: registration_year, Length: 95, dtype: int64

In [23]:
cars_before_1900 = autos["registration_year"] < 1960
cars_before_1900.value_counts()

False    48184
True        47
Name: registration_year, dtype: int64

In [24]:
cars_after_2016 = autos["registration_year"] > 2016
cars_after_2016.value_counts()

False    46364
True      1867
Name: registration_year, dtype: int64

The cars registered before 1960 only tally up to 47. However, when we increase the year parameter to 1970, the jump in cars registered rockets to 400 meaning that 1960 is good cutoff for the low end. Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. 

In [25]:
autos = autos[autos["registration_year"].between(1960,2016)]

In [26]:
autos["registration_year"].describe()

count    46317.000000
mean      2002.988643
std          6.921129
min       1960.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

Our registation dates got narrowed down to 463167 values. The mean registration year was 2002 with a standard deviation of roughly 7. This is important because it tells us that the majortiy of cars being sold on Ebay are old used cars.  

# Exploring price by Brand

In [27]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen'],
      dtype='object')

In [28]:
autos["brand"].value_counts(normalize=True)

len(autos["brand"].unique())

40

German cars represent over 50% of the cars listed on Ebay. Volkswagon being the most common vehicle with rougly 20% of the listings. 

In [29]:
brand_counts = autos["brand"].value_counts(normalize=True)
top_brands = brand_counts[brand_counts > 0.04].index
print(top_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford',
       'renault'],
      dtype='object')


The brand column consists of 40 unique car brands, all of which have a different listing share on Ebay. Using sort values it becomes clear that german cars are of the most listed vehicle brands on ebay, accounting for over 50% of the ads. Several car brands in the bottom of the list have very minor share of ebay listings.

For our aggregation purposes we will only keep the any car brands that account for over 4% of vehicle ads on the website. The brands are: 'Volkswagon', 'bmw', 'opel', merceded, audi, ford, and renault.

In [30]:
brand_mean_prices = {}

for brand in top_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices


{'volkswagen': 5636,
 'bmw': 8621,
 'opel': 3004,
 'mercedes_benz': 8609,
 'audi': 9380,
 'ford': 4050,
 'renault': 2496}

after analyzing the mean prices of the top car brands on ebay we see that renault's have the lowest prices of the bunch, while audi had the highest average prices on the ads. Volkswagon is in the middle.

# Checking Mileage by Brand

In [32]:
brand_mean_mileage = {}

for brand in top_brands: 
    specific_brand = autos[autos["brand"] == brand]
    mean_mileage = specific_brand["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
    
brand_mean_mileage
    

{'volkswagen': 128804,
 'bmw': 132723,
 'opel': 129411,
 'mercedes_benz': 131158,
 'audi': 129245,
 'ford': 124412,
 'renault': 128337}

To find out if there's any visible link between a veehicle's mileage and its price, we have derived both the top_brand's mean price and mean mileage. Furthermore, we can convert both dictionaries to series and then a datatframe. 

In [35]:
bmp_series = pd.Series(brand_mean_prices) # Using the series constructor we convert the brand price dictionary to a sereis
print(bmp_series)
bmm_series = pd.Series(brand_mean_mileage)
print(bmm_series)

volkswagen       5636
bmw              8621
opel             3004
mercedes_benz    8609
audi             9380
ford             4050
renault          2496
dtype: int64
volkswagen       128804
bmw              132723
opel             129411
mercedes_benz    131158
audi             129245
ford             124412
renault          128337
dtype: int64


Using the first series object we will create a datarame (df) and use the top_brands as the index. After that, we will add another column containing the mean mileage values.

In [46]:
brand_info = pd.DataFrame(bmp_series, columns=['mean_price']) 
brand_info

mean_price
volkswagen           5636
bmw                  8621
opel                 3004
mercedes_benz        8609
audi                 9380
ford                 4050
renault              2496

In [54]:
brand_info["mean_mileage"] = bmm_series #Adding mean_mileage series as a column in brand info
brand_info

mean_price  mean_mileage
volkswagen           5636        128804
bmw                  8621        132723
opel                 3004        129411
mercedes_benz        8609        131158
audi                 9380        129245
ford                 4050        124412
renault              2496        128337

By looking at the brand_info dataframe above we can deduce. bmw's have the highest average mileage and the second highest lsiting prices. Renaults' have the lowest pricing with second lowest avg mileage. 